### ISSUE TITLE
sqlite dataset fails to raise a StopIteration: incorrect result

### ISSUE THREAD

**Issue by t-kalinowski** [tensorflow/tensorflow#33253]
**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): Yes
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Ubuntu 18.04
- TensorFlow installed from (source or binary): issue is reproducible with both
- TensorFlow version (use command below): Confirmed on 2.0 and 1.14 
- Python version: python2

**Describe the current behavior**
After creating a `tf.data.experimental.SqlDataset()` and performing a map and batch operaion, the dataset fails to raise a StopIteration after going through the entire database, and begins to repeat / recycle values incorrectly. 

**Describe the expected behavior**
The dataset stops after returning all the records in the sqlite database.

**Code to reproduce the issue**
```python
# --- Create a dummy sqlite database ---
import os
import sqlite3
pth = "/tmp/bug-report.sqlite"
if os.path.exists(pth): os.unlink(pth)

query = 'SELECT * FROM data'

con = sqlite3.connect(pth)
c = con.cursor()
c.execute('CREATE TABLE data (col1 Int)')

for i in range(3):
  c.execute('INSERT INTO data VALUES (' + str(i) + ')')

con.commit()

# print the db, just to show what's in there
c.execute(query)
print "Actual query results: ", c.fetchall()
con.close()

# --- create a tf sqlite dataset ---
import tensorflow as tf
print tf.version.VERSION

ds = tf.data.experimental.SqlDataset('sqlite', pth, query, (tf.int32))
ds = ds.map(lambda x: tf.identity(x))


# this is supposed to terminate after only two batchs since the sqlite db only
# has 2 entries, but it goes forever
print "Batch size of 2:"
i = 0
for e in ds.batch(2):
  print e
  
  i += 1
  if i > 2: print "  Should have stopped by now"
  if i > 10: print "    breaking early"; break
  
# if batch size is larger than the db size, it also fails to stop
print "Batch size of 4:"
i = 0
for e in ds.batch(4):
  print e
  
  i += 1
  if i > 1: print "  Should have stopped by now"
  if i > 10: print "    breaking early"; break

# if batch size is exactly a multiple of the sqlite db size, then it does
# raise a StopIteration correctly
print "Batch size of 3:"
for e in ds.batch(3):
  print e
```

![image](https://user-images.githubusercontent.com/8462255/66663537-ecef8080-ec18-11e9-84fd-afd02c8a66ab.png)

**Other info / logs**
This has been reproduced on tensorflow versions 2.0, 1.14 and 1.15

**Comment by yongtang** [comment]
Added PR #33271 for the fix.

**Comment by t-kalinowski** [comment]
Thank you for the quick fix! Any chance this could make it into 1.15 too?

**Comment by yongtang** [comment]
@t-kalinowski We will have to wait for the PR to be merged into master. Once it is in the master, it might be cherry-picked into release 1.15 or 2.0. Though given 1.15's release schedule, I would not count on this fix being picked up in 1.15.
